invenio_records_permissions/permissions/base.py

```diff
@@ -6,99 +6,108 @@
-# |-------------|------|----------|--------------|
-#
+    TODO: Recognize a PermissionPolicy class in other modules instead of
+          individual factory functions to lessen the configuration burden.
+    """
 
+    # Deny all by default
+    can_list = [Deny()]
+    can_create = [Deny()]
+    can_read = [Deny()]
+    can_update = [Deny()]
+    can_delete = [Deny()]
+
+    def __init__(self, action, **over):
+        """Constructor."""
+        super(BasePermissionPolicy, self).__init__()
+        self.action = action
+        self.over = over
 
-class BasePermission(Permission):
+    @property
+    def generators(self):
+        """List of Needs generators for self.action."""
+        action = self.action
+        cls = self.__class__
+        if hasattr(cls, 'can_' + action):
+            return getattr(cls, 'can_' + action)
+
+        current_app.logger.error("Unknown action {action}.".format(
+            action=action))
 
-    def __init__(self, config, action):
-        super(BasePermission, self).__init__()
-        self.config = config
-        self.permission_list = self.config.get_permission_list(action)
+        # Question: should this default to [Deny()] or []?
+        return [Deny()]
 
     @property
     def needs(self):
-        # Needs caching cannot be done here, since sometimes depends on the
-        # record. It must be implemented in each generator.
-        needs = []
-        for needs_generator in self.permission_list:
-            tmp_need = needs_generator.needs()
-            if tmp_need:
-                needs.extend(tmp_need)
-
-        self.explicit_needs = self.explicit_needs.union(needs)
-        self._load_permissions()
+        """Set of generated Needs.
 
+        If ANY of the Needs are matched, permission is allowed.
+        """
+        needs = [
+            generator.needs(**self.over) for generator in self.generators
+        ]
+        needs = set(chain.from_iterable(needs))
+
+        self.explicit_needs |= needs
+        self._load_permissions()  # explicit_needs is used here
         return self._permissions.needs
 
     @property
     def excludes(self):
-        # Needs caching cannot be done here, since sometimes depends on the
-        # record. It must be implemented in each generator.
-        excludes = []
-        for excludes_generator in self.permission_list:
-            tmp_exclude = excludes_generator.excludes()
-            if tmp_exclude:
-                excludes.extend(tmp_exclude)
-
-        self.explicit_needs = self.explicit_needs.union(excludes)
+        """Set of excluded Needs.
+
+        NOTE: `excludes` take precedence over `needs` i.e., if the same Need
+        is in the `needs` list and the `excludes` list, then that Need is
+        excluded (must not be provided by identity).
+        """
+        excludes = [
+            generator.excludes(**self.over) for generator in self.generators
+        ]
+        excludes = set(chain.from_iterable(excludes))
```

Comment:
As soon as https://github.com/inveniosoftware/invenio-access/pull/166 gets merged, the we can use `explicit_excludes` coming from the parent class.

Something like: ``self.explicit_excludes |= excludes``

Reply:
Yes. I will make another PR for that once it is merged.